File: src/klikatech/_1_week/_3_arrays_generics/Example2.scala
package klikatech._1_week._3_arrays_generics

import scala.util.Try

// Generics
object Example2 {
    // public final java.util.List list = java.util.Arrays.asList(1, 2, 3, 4);
    val list: List[Int] = List(0, 1, 2, 3, 4)
    val elem2: Int = list(2) // Java: list.get(2)
    //val elem5: Try[Int] = Try(list(5))

    def main(args: Array[String]) {
        println("elem2: " + elem2)
        //println("elem5: " + elem5.getOrElse(println("ops!")))
    }
}

File: src/klikatech/_7_week/_1_generics/_0_java_arrays/Example2.java
package klikatech._7_week._1_generics._0_java_arrays;

/*
    Java: USE defined variance          = место использования
    Scala: DEFINITION defined variance  = место объявления

    In Java all generic types are invariant.

                Invariant       Covariant       Contra-variance
    Java        <T>             <? extends T>   <? super T>
    Scala       [T]             [+T]            [-T]
    Scala       [T]             [_ <: T]        [_ >: T]
    (wildcards)

    String <: Object
 */
public class Example2 {

    private static class MyClass<T> {
    }

    public static void main(String[] args) {
        // 1
        // non-variant
        //MyClass<String> ref11 = new MyClass<Object>();
        //MyClass<Object> ref12 = new MyClass<String>();

        MyClass<String> ref13 = new MyClass<String>();

        // 2
        // co-variant
        MyClass<? extends Object> ref2 = new MyClass<String>(); // Child

        // 3
        // contra-variant
        MyClass<? super String> ref3 = new MyClass<Object>(); // Parent

    }
}
